internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go

```diff
@@ -93,52 +115,328 @@ func (c *gitHubWatcherController) Sync(ctx controllerlib.Context) error {
+		UsernameAttribute:  usernameAttribute,
+		OAuth2Config: &oauth2.Config{
+			ClientID:     clientID,
+			ClientSecret: clientSecret,
+		},
+		AllowedOrganizations:    upstream.Spec.AllowAuthentication.Organizations.Allowed,
+		OrganizationLoginPolicy: policy,
+		AuthorizationURL:        fmt.Sprintf("%s/login/oauth/authorize", host),
+		HttpClient:              httpClient,
+	}
+	return providerConfig, errors.NewAggregate(applicationErrors)
+}
+
+func (c *gitHubWatcherController) validateHostAndTLSConfiguration(ctx context.Context, gitHubAPIConfig v1alpha1.GitHubAPIConfig) ([]*metav1.Condition, string, *http.Client, error) {
+	githubConnectionUnknown := &metav1.Condition{
+		Type:    GitHubConnectionValid,
+		Status:  metav1.ConditionUnknown,
+		Reason:  "UnableToValidate",
+		Message: "unable to validate; see other conditions for details",
+	}
+
+	tlsConfigurationUnknown := &metav1.Condition{
+		Type:    TLSConfigurationValid,
+		Status:  metav1.ConditionUnknown,
+		Reason:  "UnableToValidate",
+		Message: "unable to validate; see other conditions for details",
+	}
+
+	// Should not happen due to CRD defaulting
+	if gitHubAPIConfig.Host == nil || len(*gitHubAPIConfig.Host) < 1 {
+		return []*metav1.Condition{
+			{
+				Type:    HostValid,
+				Status:  metav1.ConditionFalse,
+				Reason:  "InvalidHost",
+				Message: `spec.githubAPI.host ("") is not valid`,
+			},
+			tlsConfigurationUnknown,
+			githubConnectionUnknown,
+		}, "", nil, nil
+	}
+
+	host := *gitHubAPIConfig.Host
+	hostPort, addressParseErr := endpointaddr.Parse(host, 443)
+	if addressParseErr != nil {
+		// addressParseErr is not recoverable. It requires user interaction, so do not return the error.
+		return []*metav1.Condition{
+			{
+				Type:    HostValid,
+				Status:  metav1.ConditionFalse,
+				Reason:  "InvalidHost",
+				Message: fmt.Sprintf("spec.githubAPI.host (%q) is not valid", host),
+			},
+			tlsConfigurationUnknown,
+			githubConnectionUnknown,
+		}, host, nil, nil
+	}
+
+	hostOnly := *gitHubAPIConfig.Host
+	hostWithHttps := fmt.Sprintf("https://%s", hostOnly)
+	conditions := make([]*metav1.Condition, 0)
+	conditions = append(conditions, &metav1.Condition{
+		Type:    HostValid,
+		Status:  metav1.ConditionTrue,
+		Reason:  upstreamwatchers.ReasonSuccess,
+		Message: fmt.Sprintf("spec.githubAPI.host (%q) is valid", host),
+	})
+
+	certPool, _, buildCertPoolErr := buildCertPool(gitHubAPIConfig.TLS)
+	if buildCertPoolErr != nil {
+		// buildCertPoolErr is not recoverable. It requires user interaction, so do not return the error.
+		conditions = append(conditions, &metav1.Condition{
+			Type:    TLSConfigurationValid,
+			Status:  metav1.ConditionFalse,
+			Reason:  "InvalidTLSConfig",
+			Message: fmt.Sprintf("tls configuration is not valid for spec.githubAPI.host (%q): %s", host, buildCertPoolErr),
+		})
+		conditions = append(conditions, githubConnectionUnknown)
+
+		return conditions, hostWithHttps, nil, nil
+	}
+
+	tlsDialer := tls.Dialer{
```

Comment:
This likely needs to be passed in & mocked in tests for IPv6, per above comment.

Reply:
we can use a real dialer, since CI supports IPv6